File: Python/Merge k Sorted Lists.py
"""
Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Divide and Conquer
class Solution:
    # @param a list of ListNode
    # @return a ListNode
    def mergeKLists(self, lists):
        if len(lists) == 1:
            return lists[0]
        if len(lists) == 0:
            return None
        while len(lists) >= 2:
            newList = []
            for i in xrange(len(lists)/2):
                newList.append(self.mergeTwoLists(lists.pop(),lists.pop(0)))
            if len(lists) > 0 :
                newList.append(lists.pop())
            lists = newList
        return lists.pop()
        
    def mergeTwoLists(self, l1, l2):
        if l1 is None:
            return l2
        if l2 is None:
            return l1
        dummy = ListNode(0)
        head = dummy
        while l1 and l2:
            if l1.val > l2.val:
                head.next = l2
                l2=l2.next
                head = head.next
            else:
                head.next = l1
                l1=l1.next
                head = head.next
        if l1:
            head.next = l1
        elif l2:
            head.next = l2
        else:
            head.next = None
        return dummy.next

# Priority Queue
class Solution:
    # @param a list of ListNode
    # @return a ListNode
    def mergeKLists(self, lists):
        dummy = ListNode(0)
        current = dummy
        
        heap = []
        for sorted_list in lists:
            if sorted_list:
                heapq.heappush(heap, (sorted_list.val, sorted_list))
                
        while heap:
            smallest = heapq.heappop(heap)[1]
            current.next = smallest
            current = current.next
            if smallest.next:
                heapq.heappush(heap, (smallest.next.val, smallest.next))
                
        return dummy.next

File: Java/Merge k Sorted Lists.java
/*
Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

//O(nk log k) runtime, O(1) space – Divide and conquer using two way merge:
public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists.length == 0) return null;
        int end = lists.length -1;
        while(end > 0)
        {
            int begin = 0;
            while(begin < end)
            {
                lists[begin] = mergeTwoLists(lists[begin],lists[end]);
                begin++;
                end--;
            }
        }
        return lists[0];
    }
    
    
    private ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode prev = dummy;
        while(l1 != null && l2 != null)
        {
            if(l1.val <= l2.val)
            {
                prev.next = l1;
                l1 = l1.next;
            }
            else
            {
                prev.next = l2;
                l2 = l2.next;
            }
            prev = prev.next;
        }
        if(l1 != null) prev.next = l1;
        if(l2 != null) prev.next = l2;
        return dummy.next;
    }
}


//O(nk log k) runtime, O(k) space – Heap:
public class Solution {
    private static final Comparator<ListNode> listComparator =
      new Comparator<ListNode>() {
        @Override
        public int compare(ListNode x, ListNode y) {
            return x.val - y.val;
        }
    };
    
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;
        Queue<ListNode> queue = new PriorityQueue<>(lists.length, listComparator); 
        for (ListNode node : lists) {
            if (node != null) {
                queue.add(node);
            } 
        }
        ListNode dummyHead = new ListNode(0);
        ListNode p = dummyHead;
        while (!queue.isEmpty()) {
            ListNode node = queue.poll();
            p.next = node;
            p = p.next;
            if (node.next != null) {
                queue.add(node.next);
            }
        }
        return dummyHead.next;
    }
}